source/sigv4.c

```diff
@@ -708,154 +729,427 @@ static SigV4Status_t getCredentialScope( SigV4Parameters_t * pSigV4Params,
+        *pNumberOfParameters = numberOfParameters;
+    }
 
-        while( tokenQueries != NULL )
-        {
-            canonicalRequest->pQueryLoc[ index ] = &tokenQueries[ 0 ];
-            tokenQueries = strtok( NULL, "&" );
+    static SigV4Status_t writeCanonicalQueryParameters( CanonicalContext_t * pCanonicalRequest,
+                                                        size_t numberOfParameters )
+    {
+        SigV4Status_t returnStatus = SigV4Success;
+        char * pBufLoc = NULL;
+        size_t encodedLen = 0U, remainingLen = 0U, i = 0U;
 
-            index++;
-        }
+        assert( pCanonicalRequest != NULL );
+        assert( pCanonicalRequest->pBufCur != NULL );
+        assert( pCanonicalRequest->pQueryLoc != NULL );
 
-        qsort( canonicalRequest->pQueryLoc, index, sizeof( char * ), cmpKeyValue );
+        pBufLoc = pCanonicalRequest->pBufCur;
+        remainingLen = pCanonicalRequest->bufRemaining;
 
-        for( i = 0U; i < index; i++ )
+        for( i = 0U; i < numberOfParameters; i++ )
         {
-            tokenParams = strtok( canonicalRequest->pQueryLoc[ i ], "=" );
-
-            if( tokenParams != NULL )
+            assert( pCanonicalRequest->pQueryLoc[ i ].key.pData != NULL );
+            assert( pCanonicalRequest->pQueryLoc[ i ].key.dataLen > 0U );
+
+            encodedLen = remainingLen;
+            returnStatus = encodeURI( pCanonicalRequest->pQueryLoc[ i ].key.pData,
+                                      pCanonicalRequest->pQueryLoc[ i ].key.dataLen,
+                                      pBufLoc,
+                                      &encodedLen,
+                                      true,
+                                      false );
+
+            if( returnStatus == SigV4Success )
             {
-                encodeURI( tokenParams, strlen( tokenParams ), pBufLoc, &remainingLen, true, false );
-                pBufLoc += remainingLen;
-                *pBufLoc = '='; /* Overwrite null character. */
-
-                canonicalRequest->bufRemaining -= remainingLen;
-                remainingLen = canonicalRequest->bufRemaining;
+                pBufLoc += encodedLen;
+                remainingLen -= encodedLen;
+                encodedLen = remainingLen;
+
+                /* An empty value corresponds to an empty string. */
+                if( ( pCanonicalRequest->pQueryLoc[ i ].value.dataLen > 0U ) &&
+                    ( pCanonicalRequest->pQueryLoc[ i ].value.pData != NULL ) )
+                {
+                    if( encodedLen < 1U )
+                    {
+                        returnStatus = SigV4InsufficientMemory;
+                    }
+                    else
+                    {
+                        *pBufLoc = '=';
+                        ++pBufLoc;
+                        remainingLen -= 1;
+                        encodedLen = remainingLen;
+                        returnStatus = encodeURI( pCanonicalRequest->pQueryLoc[ i ].value.pData,
```

Comment:
I get that `encodeURI` uses the same variable for both destination buffer length and bytes written, but that leads to code such as:
```
pBufLoc += encodedLen;
remainingLen -= encodedLen;
encodedLen = remainingLen;
```
Consider separating that parameter into two, so that `encodedLen` doesn't have to be reassigned to `remainingLen`

Reply:
I agree, was just in a rush. Might be able to fix this if time permits.